### ISSUE TITLE
[java] AppendCharacterWithChar should consider return type of method calls

### ISSUE THREAD

**Issue by Zustin** [pmd/pmd#3932]
**Affects PMD Version:** 6.42.0

**Rule:** [ArrayIsStoredDirectly](https://pmd.github.io/latest/pmd_rules_java_performance.html#appendcharacterwithchar)

**Description:**
Hi, I found a false negative about the rule AppendCharacterWithChar. PMD should have reported a warning in line 5. But it doesn't. I think this false negative is caused by the type of method return value. Thanks for your consideration.

**Code Sample demonstrating the issue:**

```java
public String getLiteral() {
    return "a";
}
public void bar(StringBuffer sb) {
    sb.append(getLiteral());  // should report a warning here
}
```

**Expected outcome:**

PMD should report a violation at line 5, but doesn't. This is a false-negative.

**Running PMD through:** *[Maven]*

**Comment by oowekyala** [comment]
I don't think this is a false negative. The rule recommends a local refactoring that can be done transparently for clients (changing a string literal to a char literal). If the rule found a violation in the code you show, then the fix would be to change the return type of a method, or create a new method which returns a char. This is not a simple local refactoring anymore, and could have unintended consequences. I think treating methods as black boxes is the right thing to do here.
